README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddResizeImage

A widget for ManagerManager plugin that allows image size to be changed (TV) so it is possible to create a little preview (thumb).

The widget activates as documents are saved (event `OnBeforeDocFormSave`) and it does not recreate images so server load is minimal.


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager) >= 0.7
* [(MODX)EvolutionCMS.snippets.ddGetMultipleField](https://code.divandesign.biz/modx/ddgetmultipleField) >= 3.6 (if mm_ddMultipleFields fields unparse is required)


## Documentation


### Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/ddresizeimage/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.biz/modx/ddmmeditor).
* [(MODX)EvolutionCMS.snippets.ddGetMultipleField](https://code.divandesign.biz/modx/ddgetmultiplefield).


### Parameters description

* `$params`
	* Desctription: Parameters, the pass-by-name style is used.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* **Required**
	
* `$params->fields`
	* Desctription: The names of TVs for which the widget is applied.
	* Valid values: `stringCommaSeparated`
	* **Required**
	
* `$params->width`
	* Desctription: Width of the image being created (in px).  
		Empty value means width calculating automatically according to height.  
		At least one of the two parameters must be defined.
	* Valid values: `integer`
	* **Required**
	
* `$params->height`
	* Desctription: Height of the image being created (in px).  
		Empty value means height calculating automatically according to width.  
		At least one of the two parameters must be defined.
	* Valid values: `integer`
	* **Required**
	
* `$params->filenameSuffix`
	* Desctription: The suffix for the images being created.  
		Its empty value makes initial images to be rewritten!
	* Valid values: `string`
	* Default value: `'_ddthumb'`
	
* `$params->transformMode`
	* Desctription: Transform mode.
	* Valid values:
		* `'resize'` — resize only, the image will be inscribed into the specified sizes with the same proportions
		* `'crop'` — crop only
		* `'resizeAndCrop'` — resize small side then crop big side to the specified value
		* `'resizeAndFill'` — inscribe image into the specified sizes and fill empty space with the specified background (see `$params->backgroundColor`)
	* Default value: `'resizeAndCrop'`
	
* `$params->backgroundColor`
	* Desctription: Result image background color in HEX (used only for `$params->transformMode` == `'resizeAndFill'`).
	* Valid values: `string`
	* Default value: `'#ffffff'`
	
* `$params->allowEnlargement`
	* Desctription: Allow output enlargement.
	* Valid values: `boolean`
	* Default value: `true`
	
* `$params->quality`
	* Desctription: JPEG compression level.
	* Valid values: `integer`
	* Default value: `$modx->getConfig('jpegQuality')`
	
* `$params->watermarkImageFullPathName`
	* Desctription: Specify if you want to overlay your image with watermark.  
		You can pass a relative path too (e. g. `assets/images/some.jpg`), the widget will automatically add `base_path` if needed.
	* Valid values: `string`
	* Default value: —
	
* `$params->replaceDocFieldVal`
	* Desctription: TV values rewriting status.  
		Note that the parameter isn't used if `$params->ddMultipleField_isUsed` == `1`!
	* Valid values:
		* `0`
		* `1` — tv values are replaced by the names of the created images
	* Default value: `0`
	
* `$params->ddMultipleField_isUsed`
	* Desctription: Multiple field status (for using with (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields).
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `$params->ddMultipleField_columnNumber`
	* Desctription: The number of the column in which the image is located (for using with (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields).
	* Valid values: `integer`
	* Default value: `0`
	
* `$params->ddMultipleField_rowNumber`
	* Desctription: The number of the row that will be processed (for using with (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields).
	* Valid values:
		* `integer`
		* `'all'`
	* Default value: `'all'`
	
* `$params->ddMultipleField_rowDelimiter`
	* Desctription: The row delimiter (for using with (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields).
	* Valid values: `string`
	* Default value: `'||'`
	
* `$params->ddMultipleField_colDelimiter`
	* Desctription: The column delimiter (for using with (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields).
	* Valid values: `string`
	* Default value: `'::'`
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[i]`
	* Desctription: CMS user role.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applying to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[i]`
	* Desctription: Templates ID.
	* Valid values: `integer`
	* **Required**


### CMS events

* `OnBeforeDocFormSave`


### Examples


#### Creating thumbnails 200 × auto px

The width must be equal to 200 px and the height must be calculated automatically.

```
mm_ddResizeImage([
	'fields' => 'imageTV',
	'width' => 200
]);
```


#### Creating thumbnails 200 × 100 px

The width must be equal to 200 px and the height must be equal to 100 px.

```
mm_ddResizeImage([
	'fields' => 'imageTV',
	'width' => 200,
	'height' => 100
]);
```


#### Overwrite TV values with resized images (`$params->replaceDocFieldVal` == `1`)

```
mm_ddResizeImage([
	'fields' => 'imageTV',
	'width' => 200,
	'height' => 100,
	'replaceDocFieldVal' => 1
]);
```

The images will uploaded and after document saving the TV values was changed fom `'assets/images/some.jpg'` to `'assets/images/some_ddthumb.jpg'`.


#### Providing the images that was uploaded to be limited to 125 × 68 px (`$params->transformMode` == `'resizeAndFill'`)

All that is uploading must be limited to the size and when proportions does not match then blank space must be filled with `#686868` color.

```
mm_ddResizeImage([
	'fields' => 'imageTV',
	'width' => 125,
	'height' => 68,
	'transformMode' => `'resizeAndFill'`,
	'filenameSuffix' => '',
	'backgroundColor' => '#686868'
]);
```


#### Overlay images with watermark (`$params->watermarkImageFullPathName`)

```
mm_ddResizeImage([
	'fields' => 'general_photo',
	'width' => 1920,
	'height' => 1080,
	'watermarkImageFullPathName' => 'assets/images/logo.png'
]);
```


## Links

* [Home page](https://code.divandesign.biz/modx/mm_ddresizeimage)
* [Telegram chat](https://t.me/dd_code)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddResizeImage changelog


## Version 1.6.1 (2021-10-02)
* \* Attention! (MODX)EvolutionCMS.snippets.ddGetMultipleField >= 3.6 is required (if mm_ddMultipleFields fields unparse is required).
* \* `\DDTools\Snippet::runSnippet` is used instead of `$modx->runSnippet` to run (MODX)EvolutionCMS.snippets.ddGetMultipleField without DB and eval.


## Version 1.6 (2020-07-29)
* \+ Added GIF support.
* \* Parameters:
	* \+ `$params->watermarkImageFullPathName`: The new parameter. Specify if you want to overlay your image with watermark.
	* \- `$params->croppingMode`: The parameter was removed (with backward compatibility).
	* \+ `$params->transformMode`: The new parameter (with backward compatibility to `$params->croppingMode`).
* \* Refactoring:
	* \* `\DDTools\FilesTools::modifyImage` is used.
	* \* `\DDTools\ObjectTools::isPropExists` is used instead of `isset`.
	* \* `\DDTools\ObjectTools::extend` is used instead of `array_merge`.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \+ Composer.json.


## Version 1.5 (2017-04-17)
* \* PHP 7.x compatibility.


## Version 1.4 (2016-11-01)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.7 is required.
* \+ Added the ability to set output image quality level (closes #1). Many thanks to @Aharito for initiative and patiance :pray:!
* \* Refactoring, the widget now using named parameters (with backward compatibility).
* \* Short array syntax is used because it's more convenient.
* \* Minor changes for compatibility with (MODX)EvolutionCMS.snippets.ddGetMultipleField 3.3.
* \* phpthumb: Fixed constructor name for compatibility with PHP7 (from 4bf5df5f14222d44de1ad78e71eedc3a8c7ade7a by @yama).


## Version 1.3.5 (2014-05-21)
* \* Minor changes for compatibility with (MODX)EvolutionCMS.snippets.ddGetMultipleField 3.x.
* \* Other minor changes.


## Version 1.3.4 (2013-07-19)
* \* The mm_ddMultipleFields fields empty check has been added.
* \* The images dimension check has been added (to exclude the situations when the input data is not an image).
* \* A suffix is not added to a file name if it already has one (it used to happen if `$replaceFieldVal` == `1`).


## Version 1.3.3 (2013-07-09)
* \* Checking for the datatype of result of function `tplUseTvs` has been added.


## Version 1.3.2 (2013-07-02)
* \* Predefined file format bug fix.


## Version 1.3.1 (2013-06-14)
* \* Strict case for `$modx->runSnippet`.


## Version 1.3 (2013-06-01)
* \+ Parameters → `allowEnlargement`: The new parameter allowing image enlargement to be forbidden.
* \* Minor changes.


## Version 1.2b (2012-08-30)
* \* The EasyPhpThumbnail library is no longer being used, it was replaced by phpThumb.
* \* Small changes in the code and errors elimination.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

ddresizeimage/ddresizeimage.php
<?php
/**
 * mm_ddResizeImage
 * @version 1.6 (2020-07-29)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.biz/modx/mm_ddresizeimage
 * 
 * @copyright 2012–2020 DD Group {@link https://DivanDesign.biz }
 */

function mm_ddResizeImage($params){
	global $modx;
	
	$e = &$modx->Event;
	
	//For backward compatibility
	if (func_num_args() > 1){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'roles',
				'templates',
				'width',
				'height',
				'transformMode',
				'filenameSuffix',
				'replaceDocFieldVal',
				'backgroundColor',
				'ddMultipleField_isUsed',
				'ddMultipleField_columnNumber',
				'ddMultipleField_rowDelimiter',
				'ddMultipleField_colDelimiter',
				'ddMultipleField_rowNumber',
				'allowEnlargement'
			]
		]);
	}
	
	//For backward compatibility
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			$params,
			\ddTools::verifyRenamedParams(
				$params,
				[
					'transformMode' => 'croppingMode'
				]
			)
		]
	]);
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
// 				'fields' => '',
				'width' => '',
				'height' => '',
				'filenameSuffix' => '_ddthumb',
				'transformMode' => 'resizeAndCrop',
				'backgroundColor' => '#FFFFFF',
				'allowEnlargement' => 1,
				'quality' => $modx->getConfig('jpegQuality'),
// 				'watermarkImageFullPathName' => '',
				'replaceDocFieldVal' => 0,
				'ddMultipleField_isUsed' => 0,
				'ddMultipleField_columnNumber' => 0,
				'ddMultipleField_rowDelimiter' => '||',
				'ddMultipleField_colDelimiter' => '::',
				'ddMultipleField_rowNumber' => 'all',
				'roles' => '',
				'templates' => ''
			],
			$params
		]
	]);
	
	//For backward compatibility
	switch ($params->transformMode){
		case '0':
			$params->transformMode = 'resize';
		break;
		
		case '1':
			$params->transformMode = 'crop';
		break;
		
		case 'crop_resized':
			$params->transformMode = 'resizeAndCrop';
		break;
		
		case 'fill_sized':
			$params->transformMode = 'resizeAndFill';
		break;
	}
	
	//Проверим, чтобы было нужное событие, чтобы были заполнены обязательные параметры и что правило подходит под роль
	if (
		$e->name == 'OnBeforeDocFormSave' &&
		//Required parameter
		\DDTools\ObjectTools::isPropExists([
			'object' => $params,
			'propName' => 'fields'
		]) &&
		(
			$params->width != '' ||
			$params->height != ''
		) &&
		useThisRule(
			$params->roles,
			$params->templates
		)
	){
		global
			$mm_current_page,
			$tmplvars
		;
		
		//Получаем необходимые tv для данного шаблона (т.к. в mm_ddMultipleFields тип может быть любой, получаем все, а не только изображения)
		$params->fields = tplUseTvs(
			$mm_current_page['template'],
			$params->fields,
			'',
			'id,name'
		);
		
		//Если что-то есть
		if (
			is_array($params->fields) &&
			count($params->fields) > 0
		){
			//Обработка параметров
			$params->replaceDocFieldVal =
				$params->replaceDocFieldVal == '1' ?
				true :
				false
			;
			$params->ddMultipleField_isUsed =
				$params->ddMultipleField_isUsed == '1' ?
				true :
				false
			;
			
			//Перебираем их
			foreach (
				$params->fields as
				$field
			){
				//Если в значении tv что-то есть
				if (
					\DDTools\ObjectTools::isPropExists([
						'object' => $tmplvars,
						'propName' => $field['id']
					]) &&
					trim($tmplvars[$field['id']][1]) != ''
				){
					$image = trim($tmplvars[$field['id']][1]);
					
					//Если это множественное поле
					if ($params->ddMultipleField_isUsed){
						//Получим массив изображений
						$images = \DDTools\Snippet::runSnippet([
							'name' => 'ddGetMultipleField',
							'params' => [
								'inputString' => $image,
								'rowDelimiter' => $params->ddMultipleField_rowDelimiter,
								'colDelimiter' => $params->ddMultipleField_colDelimiter,
								'startRow' =>
									$params->ddMultipleField_rowNumber == 'all' ?
									0 :
									$params->ddMultipleField_rowNumber
								,
								'totalRows' =>
									$params->ddMultipleField_rowNumber == 'all' ?
									'all' :
									1
								,
								'outputFormat' => 'JSON',
								'columns' => $params->ddMultipleField_columnNumber,
								//For backward compatibility with < 3.3
								'string' => $image,
								//For backward compatibility with < 3.0b
								'field' => $image,
								'splY' => $params->ddMultipleField_rowDelimiter,
								'splX' => $params->ddMultipleField_colDelimiter,
								'num' =>
									$params->ddMultipleField_rowNumber == 'all' ?
									0 :
									$params->ddMultipleField_rowNumber
								,
								'count' =>
									$params->ddMultipleField_rowNumber == 'all' ?
									'all' :
									1
								,
								'format' => 'JSON',
								'colNum' => $params->ddMultipleField_columnNumber
							]
						]);
						
						//Если пришла пустота (ни одного изображения заполнено не было)
						if (trim($images) == ''){
							$images = [];
						}elseif ($params->ddMultipleField_rowNumber == 'all'){
							$images = json_decode(
								$images,
								true
							);
						}else{
							$images = [
								trim(
									stripcslashes($images),
									'\'\"'
								)
							];
						}
					}else{
						//Запишем в массив одно изображение
						$images = [$image];
					}
					
					foreach (
						$images as
						$image
					){
						//Если есть лишний слэш в начале, убьём его
						if (strpos(
							$image,
							'/'
						) === 0){
							$image = substr(
								$image,
								1
							);
						}
						
						//На всякий случай проверим, что файл существует
						if (file_exists(
							$modx->config['base_path'] .
							$image
						)){
							//Полный путь изображения
							$imageFullPath = pathinfo(
								$modx->config['base_path'] .
								$image
							);
							
							//Если имя файла уже заканчивается на суффикс (необходимо при $params->replaceDocFieldVal == 1), не будем его добавлять
							if (
								substr(
									$imageFullPath['filename'],
									strlen($params->filenameSuffix) * -1
								) ==
								$params->filenameSuffix
							){
								$params->filenameSuffix = '';
							}
							
							//Имя нового изображения
							$newImageName =
								$imageFullPath['filename'] .
								$params->filenameSuffix .
								'.' .
								$imageFullPath['extension']
							;
							
							$modifyImageParams = (object) [
								//Ссылка на оригинальное изображение
								'sourceFullPathName' =>
									$modx->config['base_path'] .
									$image
								,
								//Формируем новое имя изображения (полный путь)
								'outputFullPathName' =>
									$imageFullPath['dirname'] .
									'/' .
									$newImageName
								,
								//Режим обрезания
								'transformMode' => $params->transformMode,
								//Ширина превьюшки
								'width' => $params->width,
								//Высота превьюшки
								'height' => $params->height,
								//Разрешить ли увеличение изображения
								'allowEnlargement' => $params->allowEnlargement,
								//Фон превьюшки (может понадобиться для заливки пустых мест)
								'backgroundColor' => $params->backgroundColor,
								//Output image quality level
								'quality' => $params->quality
							];
							
							if (\DDTools\ObjectTools::isPropExists([
								'object' => $params,
								'propName' => 'watermarkImageFullPathName'
							])){
								$modifyImageParams->watermarkImageFullPathName = $params->watermarkImageFullPathName;
							}
							
							//Делаем превьюшку
							\DDTools\FilesTools::modifyImage($modifyImageParams);
							
							//Если нужно заменить оригинальное значение TV на вновь созданное и это не $params->ddMultipleField_isUsed
							if (
								$params->replaceDocFieldVal &&
								!$params->ddMultipleField_isUsed
							){
								$tmplvars[$field['id']][1] =
									dirname($tmplvars[$field['id']][1]) .
									'/' .
									$newImageName
								;
							}
						}
					}
				}
			}
		}
	}
}
?>